tiledb/sm/filesystem/ls_scanner.h

```diff
@@ -0,0 +1,232 @@
+/**
+ * @file ls_scanner.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This defines the LsScanner class and related types used for VFS.
+ */
+
+#ifndef TILEDB_LS_SCANNER_H
+#define TILEDB_LS_SCANNER_H
+
+#include "tiledb/sm/filesystem/uri.h"
+
+#include <cstdint>
+#include <functional>
+#include <stdexcept>
+
+/** Inclusion predicate for objects collected by ls */
+template <class F>
+concept FilePredicate = true;
+
+/**
+ * DirectoryPredicate is currently unused, but is kept here for adding directory
+ * pruning support in the future.
+ */
+template <class F>
+concept DirectoryPredicate = true;
+
+namespace tiledb::sm {
+using FileFilter = std::function<bool(const std::string_view&, uint64_t)>;
+
+using DirectoryFilter = std::function<bool(const std::string_view&)>;
+/** Static DirectoryFilter used as default argument. */
+[[maybe_unused]] static bool accept_all_dirs(const std::string_view&) {
+  return true;
+}
+
+/** Type defintion for objects returned from ls_recursive. */
+using LsObjects = std::vector<std::pair<std::string, uint64_t>>;
+
+/**
+ * LsScanIterator iterates over the results of ls requests wrapped by classes
+ * deriving from LsScanner. See S3Scanner as an example.
+ *
+ * @tparam T The LsScanner type that created this iterator.
+ * @tparam U The data type stored by this iterator.
```

Comment:
Mark this comment as a point for future discussion.

It might be better to make argument `U` be an iterator itself, rather than a raw type that we're tossing around as `U*` all the time. Any change is not for this PR.